### PR TITLE
Add sdk feature into feat 2.0

### PR DESCRIPTION
**Pull request by gRoussac**
Right now this feature only build with --lib

Mainly adds some `#[cfg(feature = "sdk")]` where it is possible or not to use io/fs, or prevent sending to std with `json_pretty_print` and returning deploys 
Changes the way secret key is loaded in `with_payment_and_session()`
Publicizes some methods outside the crate 
Adds `session_bytes` to `session_executable_deploy_item()` and adds `check_no_conflicting_arg_types() `(might require additional unit tests)
Adds `with_bytes()` to `SessionStrParams `
Adds some `Clone` attributes to some structs
Removes rt-multi-thread, `The default runtime flavor is 'multi_thread', but the 'rt-multi-thread' feature is disabled.` switching (flavor = "current_thread") to tokio main

<s>I could not make it compile without this change in cargo.toml related to casper types
`getrandom = { version = "0.2.10", features = ["js"] }` getrandom probably needs to be replaced by something else.</s>
Removing `[patch.crates-io]` as I believe this bug was fixed and this is not needed anymore.

https://github.com/casper-ecosystem/rustSDK/issues/4
https://github.com/casper-ecosystem/rustSDK/issues/5

### CODE REVIEW

**Review by bpr — APPROVED** [review]
LGTM, but please wait for @Fraser999 and @zacshowa to review before merging. There's a lot of changes and I think we should tread carefully here.

File: lib/cli/deploy.rs
Comment (zacshowa) [inline lib/cli/deploy.rs:103]: The change to this function is interesting.

Perhaps we could look to overload the function by conditionally compiling 1 of 2 functions of the same name behind feature flags as opposed to modifying the body of one function for readability.

File: lib/cli/error.rs
Comment (zacshowa) [inline lib/cli/error.rs:115]: I don't think we need a `String` here. Is there a reason you need this to change in the clients lib instead of doing any needed conversion in SDK land?

File: lib/cli/parse.rs
Comment (zacshowa) [inline lib/cli/parse.rs]: I would prefer that imports stay in the order of 

```
std

external crates

crates originating from Casper

internal crate imports
```

File: lib/cli.rs
Comment (zacshowa) [inline lib/cli.rs:33]: I worry that making these public in the client, rather than re-implementing them in the SDK could create external dependencies on this code. This would cause any changes to these items to be breaking and cause it to be a headache for future maintenance of the client. Perhaps these items can just be copied over to the SDK itself.

File: src/main.rs
Comment (zacshowa) [inline src/main.rs:145]: Is there a reason we need to make the tokio runtime for the client single threaded?

File: Cargo.toml
Comment (zacshowa) [inline Cargo.toml:30]: I don't have a strong preference on how the cargo.toml is formatted. I'm just curious what lad to the changes

**Review by zacshowa — CHANGES_REQUESTED** [review]
A few things could use some changes/clarification.

File: lib/cli/error.rs
Comment (gRoussac) [reply lib/cli/error.rs:115]: Yes I modified the Error to receive a computed string. The error is dynamic, some other errors have context String too. https://github.com/casper-ecosystem/casper-client-rs/pull/106/files#diff-657c4a5be21b7e27f49860ca9e5ce63f061285609263aad05c184c1af2e12fa9R326

File: lib/cli/parse.rs
Comment (gRoussac) [reply lib/cli/parse.rs]: My VsCode is reorganizing imports on save, I am not sure about the rule applied, I guess it's alphabetic. I can try to manually re assort them.

File: lib/cli.rs
Comment (gRoussac) [reply lib/cli.rs:33]: I didn't re implement methods, every rpc call, deploys everything is relying on the casper client. The purpose is to have the client served as a library, not to duplicate code to have double maintenance, already few methods are exposed, I don't see why 2 more would be problematic. If the casper client/node evolve, then the sdk is following.

File: src/main.rs
Comment (gRoussac) [reply src/main.rs:145]: Fraser and I cleaned up imports of tokio, removing *un*necessary mutli thread rt-multi-thread to rt and the compiler asked to add that attribute.  

```
   Compiling casper-client v2.0.0 (/casper/casper-client-rs)
error: The default runtime flavor is `multi_thread`, but the `rt-multi-thread` feature is disabled.
   --> src/main.rs:145:1
    |
145 | #[tokio::main]
    | ^^^^^^^^^^^^^^
    |
    = note: this error originates in the attribute macro `tokio::main` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `casper-client` (bin "casper-client") due to previous error
```

File: Cargo.toml
Comment (gRoussac) [reply Cargo.toml:30]: I needed to compile with std and then to wasm so I needed to remove calls to io/fs and system calls. So with Fraser we tried to see what were the smallest changes. It was mostly clap and schemars giving issues. Then the rest is adding the feature it self (which is a restricting feature while a feature should actually be additive) and libs `"cdylib", "rlib"`.
Biggest change is `features = ["js"]  to getrandom`, I am not sure how to workaround that.

```shell

   Compiling getrandom v0.2.10
error: the wasm*-unknown-unknown targets are not supported by default, you may need to enable the "js" feature. For more information see: https://docs.rs/getrandom/#webassembly-support
   --> /opt/rust/.cargo/registry/src/index.crates.io-6f17d22bba15001f/getrandom-0.2.10/src/lib.rs:285:9
    |
285 | /         compile_error!("the wasm*-unknown-unknown targets are not supported by \
286 | |                         default, you may need to enable the \"js\" feature. \
287 | |                         For more information see: \
288 | |                         https://docs.rs/getrandom/#webassembly-support");
    | |________________________________________________________________________^

error[E0433]: failed to resolve: use of undeclared crate or module `imp`
   --> /opt/rust/.cargo/registry/src/index.crates.io-6f17d22bba15001f/getrandom-0.2.10/src/lib.rs:341:9
    |
341 |         imp::getrandom_inner(dest)?;
    |         ^^^ use of undeclared crate or module `imp`
```
    
The formatting changes are my VsCode with probably Better Toml extention that is making formatting on save.

File: lib/cli/deploy.rs
Comment (gRoussac) [reply lib/cli/deploy.rs:103]: Well mostly this code was repeated two times in the file and we already discussed this condition 

```
 let maybe_secret_key = if allow_unsigned_deploy && deploy_params.secret_key.is_empty() {
        None
    } else if deploy_params.secret_key.is_empty() && !allow_unsigned_deploy {
   
 ````
 
 is imo more idiomatic in if A / else if B / else. It's quite complicate to have cargo check and cargo clippy accepting overloads, I tried to have 1/2 methods with both config accordingly and that never passed cargo check/clippy. Most of times the only accepted way is to simply have #[cfg] working like if statements. Otherwise most of the times a function will be missing on compilation if you have like --features-all

File: Cargo.toml
Comment (gRoussac) [reply Cargo.toml:30]: I removed "json-schema" that was added back apparently.

Comment (gRoussac) [reply Cargo.toml:30]: I included `getrandom/js` only for sdk feature, this should be less invasive for default build.

Comment (gRoussac) [reply Cargo.toml:30]: Actually `getrandom` was removed from dev too completely already I missed that so this issue does not exist anymore.

**Review by bpr — APPROVED** [review]
LGTM

File: lib/cli/parse.rs
Comment (gRoussac) [reply lib/cli/parse.rs]: I updated the file please check it fits the reco 

https://github.com/casper-ecosystem/casper-client-rs/pull/106/commits/834032ca5646080e3d62655bd77656e666725db7

File: src/main.rs
Comment (gRoussac) [reply src/main.rs:145]: @Fraser999 Can you please confirm we want to get rid of rt-multi-thread in the client ? thanks

File: lib/cli/deploy.rs
Comment (zacshowa) [reply lib/cli/deploy.rs:103]: The structure of the if else is fine, It's functionally equivalent and should work fine.
I do think that if you like the style suggestion I put on the sister PR against the node I would love to see the same style applied to the feature flag here. If you don't want to go with that style, consider the PR approved.

**Review by zacshowa — COMMENTED** [review]
Like I said in the comment on deploy if you don't like the style suggestion consider the PR approved.

**Review by zacshowa — COMMENTED** [review]
Just trying to remove the change request now.

**Review by zacshowa — APPROVED** [review]

File: lib/cli/deploy.rs
Comment (gRoussac) [reply lib/cli/deploy.rs:103]: Casper client do not seem to have cargo clippy so I can duplicate this method in the client without side effects apparently.

Please check that commit https://github.com/casper-ecosystem/casper-client-rs/commit/4855e7ba3329eb38a960baaf2c62d5fa1121e387 thanks

**Pull request closed without merging**
